upgrade/upgrade.go

```diff
@@ -102,16 +104,27 @@ func upgrade(currentNodeID, namespace string, config *restclient.Config, lhClien
 					}
 					logrus.Infof("Finish upgrading")
 				}()
+
 				logrus.Infof("Start upgrading")
+
 				if err = doAPIVersionUpgrade(namespace, config, lhClient); err != nil {
 					return
 				}
-				if err = doPodsUpgrade(namespace, lhClient, kubeClient); err != nil {
+
+				err = retryOnConflictCause(func() error {
+					return doPodsUpgrade(namespace, lhClient, kubeClient)
+				})
+				if err != nil {
 					return
 				}
-				if err = doCRDUpgrade(namespace, lhClient); err != nil {
+
+				err = retryOnConflictCause(func() error {
+					return doCRDUpgrade(namespace, lhClient)
+				})
+				if err != nil {
```

Comment:
What if the error is not caused by the conflict? Then we will return and stuck as the leader for upgrade forever again.

Reply:
Yes, you are right. And I think that it makes sense because if the error is not caused by the conflict, something else is wrong in the upgrading process (some CRDs with new settings while other with old settings) and should stop Longhorn from working. 

If we want Longhorn managers to keep going even if there is error in one Longhorn manager, we can `defer cancel()`. I tested it yesterday. All manager will keep going even though there are errors. 

How do you think?

Reply:
`defer cancel()` is a good idea. It will not block other managers and the one always fail will end up in the leader and won't cause more harm to the system.

Reply:
But in case the Longhorn system with only 1 manager, the manger will keep going even when it fails to upgrade. I guess this is the conner case though.

Sure, added the `defer cancel()`.

Reply:
hmm? it should fail (crash) rather than keep going.

Reply:
No. It keeps going ... because there is no error returned from https://github.com/longhorn/longhorn-manager/blob/5b1bcac398d7a6b7412a56b5fbe0389f6fa9673b/upgrade/upgrade.go#L76

Reply:
Then we should fix that. Fail to pass the upgrade period should crash the process and let it retry.

Reply:
Yes, I will fix it